go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,467 @@
+}
+
+func (d *DiskCache) StoreCas(dg digest.Digest, path string) error {
+	if dg.Size > int64(d.maxCapacityBytes) {
+		return fmt.Errorf("blob size %d exceeds DiskCache capacity %d", dg.Size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg, isCas: true},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	_, exists := d.store.LoadOrStore(it.key, it)
+	if exists {
+		return nil
+	}
+	d.mu.Lock()
+	heap.Push(d.queue, it)
+	d.mu.Unlock()
+	if err := copyFile(path, d.getPath(it.key), dg.Size); err != nil {
+		return err
+	}
+	newSize := uint64(atomic.AddInt64(&d.stats.TotalSizeBytes, dg.Size))
+	atomic.AddInt64(&d.stats.TotalNumFiles, 1)
+	atomic.AddInt64(&d.stats.NumFilesStored, 1)
+	atomic.AddInt64(&d.stats.TotalStoredBytes, dg.Size)
+	if newSize > d.maxCapacityBytes {
+		select {
+		case d.gcReq <- atomic.AddUint64(&d.gcTick, 1):
+		default:
+		}
+	}
+	return nil
+}
+
+func (d *DiskCache) StoreActionCache(dg digest.Digest, ar *repb.ActionResult) error {
+	bytes, err := proto.Marshal(ar)
+	if err != nil {
+		return err
+	}
+	size := uint64(len(bytes))
+	if size > d.maxCapacityBytes {
+		return fmt.Errorf("message size %d exceeds DiskCache capacity %d", size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg, isCas: false},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	d.store.Store(it.key, it)
+
+	d.mu.Lock()
+	heap.Push(d.queue, it)
+	d.mu.Unlock()
+	if err := os.WriteFile(d.getPath(it.key), bytes, 0644); err != nil {
+		return err
+	}
+	newSize := uint64(atomic.AddInt64(&d.stats.TotalSizeBytes, int64(size)))
+	atomic.AddInt64(&d.stats.TotalNumFiles, 1)
+	atomic.AddInt64(&d.stats.NumFilesStored, 1)
+	atomic.AddInt64(&d.stats.TotalStoredBytes, int64(size))
+	if newSize > d.maxCapacityBytes {
+		select {
+		case d.gcReq <- atomic.AddUint64(&d.gcTick, 1):
+		default:
+		}
+	}
+	return nil
+}
+
+func (d *DiskCache) gc() {
+	defer func() { d.gcDone <- true }()
+	for {
+		select {
+		case <-d.shutdown:
+			return
+		case <-d.ctx.Done():
+			return
+		case <-d.gcReq:
+			start := time.Now()
+			// Evict old entries until total size is below cap.
+			for uint64(atomic.LoadInt64(&d.stats.TotalSizeBytes)) > d.maxCapacityBytes {
+				d.mu.Lock()
+				it := heap.Pop(d.queue).(*qitem)
+				d.mu.Unlock()
+				size, err := d.getItemSize(it.key)
+				if err != nil {
+					log.Errorf("error getting item size for %v: %w", it.key, err)
+					size = 0
+				}
+				atomic.AddInt64(&d.stats.TotalSizeBytes, -size)
+				atomic.AddInt64(&d.stats.TotalNumFiles, -1)
+				atomic.AddInt64(&d.stats.TotalGCedSizeBytes, size)
+				atomic.AddInt64(&d.stats.NumFilesGCed, 1)
+				it.mu.Lock()
+				diskOpsStart := time.Now()
+				// We only delete the files, and not the prefix directories, because the prefixes are not worth worrying about.
+				if err := os.Remove(d.getPath(it.key)); err != nil {
+					log.Errorf("Error removing file: %w", err)
```

Comment:
It'd be good if the error would be accumulated, so that Shutdown() could be renamed to Close() and return the accumulated errors.